jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/queue/QueueStatistics.kt

```diff
@@ -0,0 +1,227 @@
+     */
+    fun dropped(now: Long) {
+        totalPacketsDropped.increment()
+    }
+
+    companion object {
+        /**
+         * The scale to use for [RateStatistics]. 1000 means units
+         * (e.g. packets) per second.
+         */
+        private const val SCALE = 1000
+
+        /**
+         * The interval for which to calculate rates in milliseconds.
+         */
+        private const val INTERVAL_MS = 5000
+
+        /**
+         * Calculate the statistics buckets for a given queue length
+         */
+        private fun lengthBucketSizes(size: Int): LongArray {
+            val list = ArrayList<Long>()
+            list.add(0L)
+            var i = 1L
+
+            while (i < size) {
+                list.add(i)
+                i *= 4
+            }
+            val half = (size / 2).toLong()
+            if (half > list.last()) {
+                list.add(half)
+            }
+
+            val threeQuarters = (size * 3 / 4).toLong()
+            if (threeQuarters > list.last()) {
+                list.add(threeQuarters)
+            }
+
+            return list.toLongArray()
+        }
+
+        /**
+         * The queue waiting time bucket sizes.
+         */
+        private val waitBucketSizes = longArrayOf(2, 5, 20, 50, 200, 500, 1000)
+    }
+}
+
+class QueueStatistics(
+    /**
+     * The queue being monitored
+     */
+    private val queue: PacketQueue<*>
+) : PacketQueue.Observer {
+    /**
+     * A map of the time when objects were put in the queue
+     */
+    private val insertionTime = if (trackTimes) Collections.synchronizedMap(IdentityHashMap<Any, Long>()) else null
+
+    private val specific = if (debug) QueueStatisticsInternal(queue.capacity()) else null
+    private val global = globalStatsFor(queue)
+
+    override fun added(o: Any) {
+        val now = System.currentTimeMillis()
+        insertionTime?.put(o, now)
+
+        specific?.added(now)
+        global.added(now)
+    }
+
+    /**
+     * Registers the removal of a packet.
+     */
+    override fun removed(o: Any) {
+        val now = System.currentTimeMillis()
+        val queueLength = queue.size()
+        val then = insertionTime?.remove(o)
+        val wait = if (then != null) { now - then } else null
+
+        specific?.removed(now, queueLength, wait)
+        global.removed(now, queueLength, wait)
+    }
+
+    /**
+     * Registers that a packet was dropped.
+     */
+    override fun dropped(o: Any) {
+        val now = System.currentTimeMillis()
+        insertionTime?.remove(o) /* TODO: track this time in stats? */
+
+        specific?.dropped(now)
+        global.dropped(now)
+    }
+
+    val stats: OrderedJsonObject?
+        get() = specific?.stats
+
+    val queueDebugState: JSONObject
+        get() {
+            val d = queue.debugState
+            stats?.let { d["statistics"] = it }
+            return d
+        }
+
+    companion object {
+        var debug = false
+        var trackTimes = false
```

Comment:
I know these aren't constants, but when reading the code it still kinda feels like maybe they should be in all caps.  I doubt we're consistent on that though (I'm probably not even consistent myself)

Reply:
I did this, but it makes setters in Java look like `QueueObserver.Companion.setDEBUG(enabled);` which is pretty ugly.

Reply:
hm, can you use `@JvmField` here so it's just `QueueObserver.Companion.DEBUG = enabled`?

Reply:
I can, but it triggers Spotbugs complaining 
```
[ERROR] org.jitsi.utils.queue.QueueObserver.DEBUG isn't final but should be [org.jitsi.utils.queue.QueueObserver] In QueueObserver.kt MS_SHOULD_BE_FINAL
[ERROR] org.jitsi.utils.queue.QueueObserver.TRACK_TIMES isn't final but should be [org.jitsi.utils.queue.QueueObserver] In QueueObserver.kt MS_SHOULD_BE_FINAL
```

I'm inclined to just change them back to lower case.

Reply:
Ugh.  I don't feel like we should have spotbugs doing 'linting' for naming conventions like this...we've already got ktlint for that (which presumably doesn't complain?)

Reply:
It's not enforcing naming conventions -- it's enforcing "don't have static public non-final fields."

(It's safer for primitive types than it is for objects, and boolean is basically completely safe if you don't need synchronization; but it looks like their rule doesn't have an exception for it.)

Reply:
If so, then it's not complaining about the case, it's complaining about `@JvmField`, no?  Either way, this seems like a reasonable exception.  We could disable the rule for this line/scope.  Part of the problem with SpotBugs is that it's designed for Java, not Kotlin.